In [3]:
import pandas as pd

In [5]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'

In [7]:
!/opt/homebrew/bin/wget $data

--2022-09-25 13:06:33--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘data.csv’

data.csv            100%[===================>]   1.41M  4.56MB/s    in 0.3s    

2022-09-25 13:06:33 (4.56 MB/s) - ‘data.csv’ saved [1475504/1475504]



In [9]:
df = pd.read_csv('data.csv')

In [10]:
df


Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

In [10]:
import numpy as np

# Question 1

What's the version of NumPy that you installed?

You can get the version information using the __version__ field:

In [11]:
np.__version__

'1.20.3'

# Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

    16
    6572
    11914
    18990

Answer: 11914 

In [13]:
df.index

RangeIndex(start=0, stop=11914, step=1)

# Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

    Chevrolet, Volkswagen, Toyota
    Chevrolet, Ford, Toyota
    Ford, Volkswagen, Toyota
    Chevrolet, Ford, Volkswagen

    Note: You should rely on "Make" column in this question.

Answer: Chevrolet, Ford, Volkswagen

In [12]:
df['Make'].value_counts().head()

Chevrolet     1123
Ford           881
Volkswagen     809
Toyota         746
Dodge          626
Name: Make, dtype: int64

# Question 4

What's the number of unique Audi car models in the dataset?

    3
    16
    26
    34

Answer: 34

In [14]:
df[df['Make'] == 'Audi']['Model'].nunique()

34

# Question 5

How many columns in the dataset have missing values?

    5
    6
    7
    8

Answer: 5

In [36]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [16]:
# to calculate in one line:
(df.isnull().sum() != 0).sum()


5

# Question 6

    Find the median value of "Engine Cylinders" column in the dataset.
    Next, calculate the most frequent value of the same "Engine Cylinders".
    Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the    previous step.
    Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

    Hint: refer to existing mode and median functions to complete the task.

    Yes
    No

Answer: No

In [46]:
df["Engine Cylinders"].mode()

0    4.0
dtype: float64

In [53]:
df["Engine Cylinders"].median()

6.0

In [73]:
df["Engine Cylinders"].fillna(4.0).median()

6.0

# Question 7

    Select all the "Lotus" cars from the dataset.
    Select only columns "Engine HP", "Engine Cylinders".
    Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
    Get the underlying NumPy array. Let's call it X.
    Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    Compute the inverse of XTX.
    Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
    Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    What's the value of the first element of w?

    Note: You just implemented linear regression. We'll talk about it in the next lesson.

    -0.0723
    4.5949
    31.6537
    63.5643

Answer: 4.5949

In [103]:
lotus_cars = df[df.Make == "Lotus"]

In [108]:
lotus_cars = lotus_cars[['Engine HP', 'Engine Cylinders']].drop_duplicates()

In [109]:
lotus_cars

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [110]:
X = lotus_cars.values

In [111]:
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [112]:
XTX = X.T.dot(X)

In [113]:
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [115]:
XTX_inv = np.linalg.inv(XTX)

In [116]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [120]:
w = XTX_inv.dot(X.T).dot(y)

In [121]:
w

array([  4.59494481, -63.56432501])